Introduce RedactedData

jmhooper opened the pull request:
Ruby 3.2 introduced [Data](https://docs.ruby-lang.org/en/3.2/Data.html). This commit adds support for `RedactedData` where `Data` is available.

jmhooper committed:
Introduce RedactedData

Ruby 3.2 introduced [Data](https://docs.ruby-lang.org/en/3.2/Data.html). This commit adds support for `RedactedData` where `Data` is available.

zachmargolis commented on lib/redacted_struct.rb:
```diff
@@ -57,4 +51,74 @@ def pretty_print(q)
+  def base_name_for_inspection
+    if is_a? Struct
+      "struct"
+    elsif is_a? Data
```
Since we only have 2 parent classes for now, and `Data` may not be defined, let's just make it a plain `else`?
```suggestion
    else
```

zachmargolis commented on lib/redacted_struct.rb:
```diff
@@ -57,4 +51,74 @@ def pretty_print(q)
+  def base_type_for_pp
+    if is_a? Struct
+      Struct
+    elsif is_a? Data
```
ditto about the reference
```suggestion
    else
```

zachmargolis commented on lib/redacted_struct.rb:
```diff
@@ -57,4 +51,74 @@ def pretty_print(q)
+class RedactedStruct < Struct
+  include Redactable
+
+  VERSION = "1.1.0"
```
we definitely want to bump this. I'd say as big as `2.0.0` if we want!

jmhooper commented on lib/redacted_struct.rb:
```diff
@@ -1,47 +1,41 @@
     attributes = members.map do |member|
       if allowed_members.include?(member)
-        "#{member}=#{self[member].inspect}"
+        "#{member}=#{send(member).inspect}"
```
This is an unsafe send. I couldn't find a good way to avoid it without changing the API of Data. An alternative is to convert to a hash and call `#[]` on the hash.

zachmargolis replied:
yeah maybe we convert it once to hash outside the loop?

```ruby
    members_h = to_h

    attributes = members.map do |member|
      if allowed_members.include?(member)
        "#{member}=#{members_h[member].inspect}"
```

zachmargolis commented on lib/redacted_struct.rb:
```diff
@@ -1,47 +1,41 @@
+# A module that adds redacted functionality to a class
+module Redactable
+  def self.included(mod)
+    mod.extend RedactableClassMethods
```
since there's only one method being defined, WDYT about just inlining this to reduce indirection?
```suggestion
    class << mod
      attr_reader :allowed_members
    end
```

jmhooper replied:
TIL about this trick. I'll do it

zachmargolis commented on lib/redacted_struct.rb:
```diff
@@ -1,47 +1,41 @@
-  def allowed_members
-    self.class.allowed_members
+# A module that adds redacted functionality to a class
+module Redactable
```
I think it might be easier to find things if we broke these into separate files?

- `redacted_struct.rb`
- `redactable.rb`
- `redacted_data.rb`

and this one (`redacted_struct.rb`) would automatically include the other two

jmhooper replied:
I'm up for that! Willfix

zachmargolis reviewed (approved):
LGTM! Let's make sure to do some of the "release prework" and update Changelog.md as well as make  sure we bump the `version.rb`

jmhooper committed:
PR feedback

jmhooper commented:
Alright, I just responded to all the PR feedback! PTAL!

zachmargolis commented on lib/redactable.rb:
```diff
@@ -0,0 +1,70 @@
+  def inspect
+    name_or_nil = self.class.name ? " #{self.class.name}" : nil
+
+    attributes = members.map do |member|
+      members_h = to_h
```
We probably want this `to_h` on the outside so we're not re-creating it for every single key
```suggestion
    members_h = to_h

    attributes = members.map do |member|
```

zachmargolis reviewed (approved):
LGTM! 🚀

jmhooper committed:
cleanup

zachmargolis committed:
Fix typo in class documentation

- Also add ignore for rubocop Style/Documentation, it gets
  tricked by the trailing if modifier for the class definition

jmhooper committed:
update the Readme

jmhooper committed:
Merge branch 'jmhooper-add-redacted-data' of github.com:zachmargolis/redacted_struct into jmhooper-add-redacted-data

jmhooper committed:
Check if data is defined and defines define

jmhooper committed:
do not run redacted data specs if there is not redacted data

zachmargolis committed:
Update CircleCI to run a matrix build for multiple ruby versions

Hat tip: https://mikebian.co/running-tests-against-multiple-ruby-versions-using-circleci/

zachmargolis committed:
Update CircleCI

zachmargolis committed:
Revert "Update CircleCI"

This reverts commit 548f2d5fb1342c1c2b9b3028ecbd4c36841d6920.

zachmargolis committed:
CircleCI: bundler platform:

zachmargolis committed:
CircleCI orb requires rspec_junit_formatter

zachmargolis committed:
Add explanatory comment

zachmargolis committed:
Fix rubocop lint errors

zachmargolis committed:
Lint fix for 2.6

zachmargolis commented:
Thank you! I have merged, tagged the version and [pushed to rubygems.org](https://rubygems.org/gems/redacted_struct)
